DataSource/DataSource+UITableViewDelegate.swift

```diff
@@ -1,80 +1,96 @@
-//
-//  DataSource+UITableViewDelegate.swift
-//  DataSource
-//
-//  Created by Matthias Buchetics on 21/02/2017.
-//  Copyright © 2017 aaa - all about apps GmbH. All rights reserved.
-//
-
 import Differ
 import UIKit
 
 extension DataSource: UITableViewDelegate {
     // MARK: Highlighting
-    
+
     public func tableView(_ tableView: UITableView, shouldHighlightRowAt indexPath: IndexPath) -> Bool {
         let cellDescriptor = self.cellDescriptor(at: indexPath)
-        
-        if let closure = cellDescriptor.shouldHighlightClosure ?? shouldHighlight {
-            return closure(visibleRow(at: indexPath), indexPath)
+
+        if let closure = cellDescriptor?.shouldHighlightClosure ?? shouldHighlight {
+            guard let visibleRow = visibleRow(at: indexPath) else {
+                return false
+            }
+
+            return closure(visibleRow, indexPath)
         }
-        
+
         return fallbackDelegate?.tableView?(tableView, shouldHighlightRowAt: indexPath)
             ?? true
     }
-    
+
     public func tableView(_ tableView: UITableView, didHighlightRowAt indexPath: IndexPath) {
         let cellDescriptor = self.cellDescriptor(at: indexPath)
-        
-        if let closure = cellDescriptor.didHighlightClosure ?? didHighlight {
-            closure(visibleRow(at: indexPath), indexPath)
+
+        if let closure = cellDescriptor?.didHighlightClosure ?? didHighlight {
+            guard let visibleRow = visibleRow(at: indexPath) else {
+                return
+            }
+
+            closure(visibleRow, indexPath)
             return
         }
-        
+
         fallbackDelegate?.tableView?(tableView, didHighlightRowAt: indexPath)
     }
-    
+
     public func tableView(_ tableView: UITableView, didUnhighlightRowAt indexPath: IndexPath) {
         let cellDescriptor = self.cellDescriptor(at: indexPath)
-        
-        if let closure = cellDescriptor.didUnhighlightClosure ?? didUnhighlight {
-            closure(visibleRow(at: indexPath), indexPath)
+
+        if let closure = cellDescriptor?.didUnhighlightClosure ?? didUnhighlight {
+            guard let visibleRow = visibleRow(at: indexPath) else {
+                return
+            }
+
+            closure(visibleRow, indexPath)
             return
         }
-        
+
         fallbackDelegate?.tableView?(tableView, didUnhighlightRowAt: indexPath)
     }
-    
+
     // MARK: Selection
-    
+
     public func tableView(_ tableView: UITableView, willSelectRowAt indexPath: IndexPath) -> IndexPath? {
         let cellDescriptor = self.cellDescriptor(at: indexPath)
-        
-        if let closure = cellDescriptor.willSelectClosure ?? willSelect {
-            return closure(visibleRow(at: indexPath), indexPath)
+
+        if let closure = cellDescriptor?.willSelectClosure ?? willSelect {
+            guard let visibleRow = visibleRow(at: indexPath) else {
+                return indexPath
```

Comment:
ich würd hier nil zurück liefern. nil bedeutet keine selection durchführen. In dem fall is irgendwas komisch (tableview called delegate mit einem index dens gar nicht gibt)